File: app/src/main/java/me/magnum/melonds/model/InputConfig.java
package me.magnum.melonds.model;

import androidx.annotation.NonNull;

public class InputConfig {
    public static final int KEY_NOT_SET = -1;

    private Input input;
    private int key;

    public InputConfig(Input input, int key) {
        this.input = input;
        this.key = key;
    }

    public InputConfig(Input input) {
        this(input, KEY_NOT_SET);
    }

    public Input getInput() {
        return input;
    }

    public int getKey() {
        return key;
    }

    public void setKey(int key) {
        this.key = key;
    }

    public boolean hasKeyAssigned() {
        return key != KEY_NOT_SET;
    }

    @NonNull
    @Override
    public InputConfig clone() {
        return new InputConfig(this.input, this.key);
    }
}

File: app/src/main/java/me/magnum/melonds/domain/model/InputConfig.kt
package me.magnum.melonds.domain.model

data class InputConfig(
    val input: Input,
    val assignment: Assignment = Assignment.None,
    val altAssignment: Assignment = Assignment.None,
) {

    sealed class Assignment(open val deviceId: Int?) {
        data object None : Assignment(null)
        data class Key(override val deviceId: Int?, val keyCode: Int) : Assignment(deviceId)
        data class Axis(override val deviceId: Int?, val axisCode: Int, val direction: Direction) : Assignment(deviceId) {
            enum class Direction {
                POSITIVE,
                NEGATIVE,
            }
        }
    }

    fun hasKeyAssigned(): Boolean {
        return assignment != Assignment.None || altAssignment != Assignment.None
    }
}